src/tree/ActionTreeItem.ts

```diff
@@ -0,0 +1,100 @@
+    head_commit: { message: string };
+    url: string;
+    html_url: string;
+    rerun_url: string;
+    cancel_url: string;
+};
+
+export class ActionTreeItem extends AzureTreeItem {
+
+    public static contextValue: string = 'azureStaticAction';
+    public readonly contextValue: string = ActionTreeItem.contextValue;
+    public parent: ActionsTreeItem;
+    public data: GitHubAction;
+
+    constructor(parent: ActionsTreeItem, data: GitHubAction) {
+        super(parent);
+        this.data = data;
+    }
+
+    public get iconPath(): TreeItemIconPath {
+        return this.data.conclusion ? treeUtils.getThemedIconPath(path.join('conclusions', this.data.conclusion)) : treeUtils.getThemedIconPath(path.join('statuses', this.data.status));
+    }
+
+    public get id(): string {
+        return `${this.parent.parent.id}/${this.data.id}`;
+    }
+
+    public get name(): string {
+        return this.data.head_commit.message;
+    }
+
+    public get label(): string {
+        return this.name;
+    }
+
+    public get description(): string {
+        return this.data.event;
+    }
+
+    public async refreshImpl(): Promise<void> {
+        const gitHubRequest: gitHubWebResource = await createGitHubRequestOptions(undefined, this.data.url);
+        const githubResponse: IncomingMessage & { body: string } = await requestUtils.sendRequest(gitHubRequest);
+        this.data = <GitHubAction>JSON.parse(githubResponse.body);
+    }
+
+    public async rerunAction(context: IGitHubContext): Promise<void> {
+        const gitHubRequest: gitHubWebResource = await createGitHubRequestOptions(context, this.data.rerun_url, 'POST');
+        const rerunRunning: string = localize('rerunRunning', 'Rerun for action "{0}" has started.', this.data.id);
+        ext.outputChannel.appendLog(rerunRunning);
+
+        await this.waitForRunToFinish(gitHubRequest);
+        if (this.data.conclusion !== 'cancelled') {
+            const rerunCompleted: string = localize('rerunCompleted', 'Rerun for action "{0}" has completed.', this.data.id);
+            ext.outputChannel.appendLog(rerunCompleted);
+        }
+    }
+
+    public async cancelAction(context: IGitHubContext): Promise<void> {
+        const gitHubRequest: gitHubWebResource = await createGitHubRequestOptions(context, this.data.cancel_url, 'POST');
+        const cancelRunning: string = localize('cancelRunning', 'Cancel for action "{0}" has started.', this.data.id);
+        ext.outputChannel.appendLog(cancelRunning);
+
+        await this.waitForRunToFinish(gitHubRequest);
+        const cancelCompleted: string = localize('cancelCompleted', 'Cancel for action "{0}" has completed.', this.data.id);
+        ext.outputChannel.appendLog(cancelCompleted);
+    }
+
+    private async waitForRunToFinish(gitHubRequest: gitHubWebResource): Promise<void> {
+        await requestUtils.sendRequest(gitHubRequest);
+        await this.refresh(); // need to refresh to update the data
```

Comment:
> I thought about having a timeout for the refreshing of the treeItem while we wait for the action to complete. But I don't think there's any reasonable guess as to how long a rerun will take since that's entirely dependent on their project.

I'm not sure what the answer is, but I think we need a better solution here. I almost feel like you should pull out the auto-refresh logic, so that we can merge the rest of the code and have more time to figure out a solution to this separately. A few reasons I don't like this:
1. Every 2 seconds seems too often to me, since an action could take so long to complete (i.e. our Cosmos nightly build can easily take an hour)
1. The class is re-created if the user refreshes the whole tree, meaning you would continue to refresh in an instance that's no longer even displayed
1. You refresh the tree (causing at least a blip in the UI), even if nothing changes

Reply:
Okay, alternatively, I could just keep polling in the background (sort of like how like what's done [here](https://github.com/microsoft/vscode-azurestaticwebapps/blob/master/src/tree/StaticWebAppTreeItem.ts#L126-L154)) and update it once the conclusion populates.

While I think 2 seconds is kind of frequent, I wanted it to feel pretty responsive.  Maybe we could do a incremental increase in polling time?

Reply:
Can you clarify what you would do differently based on that example? The biggest difference I see is that function stops polling after a minute

Reply:
The biggest difference is that the polling is happening independently of the tree.  Right now I'm leveraging the refresh command which is what makes it look super sloppy.

This logic is similar to the `redeploy` where we await `waitForDeploymentToComplete` (which is polling the deployment logs) and then update the tree when it's finished.

Reply:
Sounds like that could help, but I'd have to wait and see what you actually do for this case

Reply:
Should I just make it in another PR?